File: swift/A11ySwiftScan/Sources/A11ySwiftScanCore/Finding.swift
import Foundation

/// A single static accessibility finding. The shape is the engine↔TS contract:
/// each item serializes to EXACTLY the JSON object the TS `scanSwift` collector
/// parses (see `src/collect-swift.ts`):
///
///     { "file": string, "line": number,
///       "ruleId": "swiftui/image-no-label" | "swiftui/control-no-name",
///       "message": string, "wcag": ["1.1.1"], "severity": "serious" | "critical" }
///
/// `wcag` is an array (always one element here) to match the web tool's
/// `Finding.wcag: readonly string[]`. `severity` is "critical" only when the
/// missing name leaves an interactive element with no accessible name at all
/// (the element is unusable by VoiceOver), else "serious".
public struct Finding: Encodable {
    public let file: String
    public let line: Int
    public let ruleId: String
    public let message: String
    public let wcag: [String]
    public let severity: String
}

File: kotlin/A11yKotlinScan/src/main/kotlin/binclusive/a11ykotlinscan/Finding.kt
package binclusive.a11ykotlinscan

/**
 * One static accessibility finding — the exact JSON shape the TS side
 * (src/collect-android-kotlin.ts) parses at the process boundary, mirroring
 * swift/A11ySwiftScan's `Finding.swift`:
 *   { file, line, ruleId, message, wcag: ["4.1.2"], severity: "serious"|"critical" }
 *
 * `severity` is "critical" only when the missing name leaves an interactive element
 * with no accessible name at all (unusable by TalkBack), else "serious".
 */
data class Finding(
    val file: String,
    val line: Int,
    val ruleId: String,
    val message: String,
    val wcag: List<String>,
    val severity: String,
)

private fun jsonString(s: String): String {
    val sb = StringBuilder("\"")
    for (c in s) {
        when (c) {
            '\\' -> sb.append("\\\\")
            '"' -> sb.append("\\\"")
            '\n' -> sb.append("\\n")
            '\r' -> sb.append("\\r")
            '\t' -> sb.append("\\t")
            else -> if (c < ' ') sb.append("\\u%04x".format(c.code)) else sb.append(c)
        }
    }
    sb.append("\"")
    return sb.toString()
}

/** Serialize findings as the JSON array the TS collector reads from stdout. Hand-rolled
 * (no kotlinx.serialization dependency) — the shape is small and fixed. */
fun findingsToJson(findings: List<Finding>): String =
    findings.joinToString(separator = ",", prefix = "[", postfix = "]") { f ->
        buildString {
            append("{")
            append("\"file\":").append(jsonString(f.file)).append(",")
            append("\"line\":").append(f.line).append(",")
            append("\"ruleId\":").append(jsonString(f.ruleId)).append(",")
            append("\"message\":").append(jsonString(f.message)).append(",")
            append("\"wcag\":").append(f.wcag.joinToString(",", "[", "]") { jsonString(it) }).append(",")
            append("\"severity\":").append(jsonString(f.severity))
            append("}")
        }
    }
